### ISSUE TITLE
Inexact adjoint of tomography operator 

### ISSUE THREAD

**Issue by tachella** [deepinv/deepinv#511]
The `A_adjoint` of the `deepinv.physics.Tomography` operator is not exact, and can cause problems when using optimization-based solvers.
I propose to replace it with the autograd-based alternative `deepinv.physics.adjoint_function` by default, and still give the opportunity to use the non-exact version.

Some examples from Nelly Pustelnik on the difference between the two operators:

when using just gradient-descent based methods (red is autograd-based, blue is current implementation)
![Image](https://github.com/user-attachments/assets/dbd3c40a-b86a-4a44-bd5e-6dd079a2a620)

when using prox-based methods
![Image](https://github.com/user-attachments/assets/29ce0042-5295-45dc-9742-0ee31145a7e9)

TODO: check run time of both alternatives

**Comment by tachella** [comment]
temporal fix for those interested

```
physics = dinv.physics.Tomography(...)
physics.A_adjoint = dinv.physics.adjoint_function(physics.A, image_size)
```

**Comment by johertrich** [comment]
Hi Julian, just as a comment since we came across exactly the same issue: the operator `physics.A` is only differentiable once (because it uses `torch.griddata`). So if we define the adjoint with `deepinv.physics.adjoint_function`, the adjoint is no longer differentiable and it throws an exception if we try to use it within autograd (e.g. for unrolling/bilevel).

My temporary fix for that is to overwrite the backward pass of the adjoint again with the forward operator itself (probably there is a shorter way of implementing it via backward hooks...):

```
class BackpropAdjoint(torch.autograd.Function):
    @staticmethod
    def forward(
        y,
        physics,
    ):
        A_adjoint = deepinv.physics.adjoint_function(
            physics.A, (y.shape[0], 1, physics.img_width, physics.img_width), device=y.device
        )
        out=A_adjoint(y)
        return out

    @staticmethod
    def setup_context(ctx, inputs, output):
        ctx.physics = inputs[1]

    @staticmethod
    def backward(ctx, grad_output):
        return ctx.physics.A(grad_output), None

physics = deepinv.physics.Tomography(...)
physics.A_adjoint = lambda y: BackpropAdjoint.apply(y, physics)
```

**Comment by tachella** [comment]
Hi Johannes, many thanks for the idea!

Maybe we could modify `deepinv.physics.adjoint_function` as you did to always rely on A for backpropagation?

**Comment by jscanvic** [comment]
It seems to be doable without too much hassle (see #519). Let me know what you think of it!

**Comment by johertrich** [comment]
Hi,

we looked a bit more into it and on one side, we agree that the current state can produce weird results.
For instance, currently the following two ways of computing the gradient of the data fidelity term are not doing the same because they are using different implementations of the adjoint.
```
physics = dinv.physics.Tomography(...)
data_fidelity = dinv.optim.L2()

x = ...
y = physics(x)
data_grad1 = data_fidelity.grad(x, y, physics)
x.requires_grad_(True)
data_grad2 = torch.autograd.grad(data_fidelity(x,y,physics),x)[0]
``` 

One the other side, I don't think that `dinv.physics.adjoint_function` will fix the issue. In the torch documentation, there is a note [here](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html) that the `grid_sample` function is nondeterministic in the backward pass on cuda. @alexdenker tried it for the deep image prior and indeed for different runs one gets completely different results. Therefore, setting this as an adjoint might be weird as well (and currently backproping through the tomography operator might induce strange behaviour). As far as I understand this is due to cuda and does not appear on the cpu.


In total I see two possible ways of making the behaviour consistent and deterministic:

- either we fully use the mismatched adjoint, in which case the backward pass of physics.A should be overwritten as physics.A_adjoint. As far as I know this is what packages like Astra do, since backprop might use an excessive amount of memory for many angles/3D (which might not be that much the scope of deepinv, but it is definitely the scope of astra).

- in the other case, we could try to replace the `grid_sample` function by rotating the image and using `torch.sum` as a reduction. Implementing and testing this might be a bit of work though.

**Comment by tachella** [comment]
Thanks for spotting this! I didn't know that `grid_sample` is not deterministic on the backward pass.

I think your first solution seems the easiest one, I guess we could still use deepinv.physics.adjoint_function but on A instead of `A_adjoint`? I'm slightly worried that the measurements will change significantly wrt to the current ones, and the changes won't be backwards compatible

**Comment by johertrich** [comment]
Yes, the first solution is very easy, I can do a PR if you like (should just be adjusting a couple of lines).  What exactly do you mean by using the adjoint_function on A instead of A_adjoint? 

I tested a bit the different variants (just comparing the two different adjoints for random images). For most setups, the differences were very small (cosine similarity > 0.998), if I take images from a CT dataset (not random images), the cosine similarity is usually >0.9999.  I would not expect that the different adjoints matter in these cases.

However, there seems to be a problem with the argument `circle=True`. The `physics.A_adjoint` cuts off everything outside the circle, while the backprop-adjoint doesn't. This is definitely a bug, but it depends on what the `circle` option should do whether the bug is in the forward or the adjoint of the operator. Can you double-check with Nelly, whether she was using the `circle=True` option?

**Comment by johertrich** [comment]
Hi, I added a pull request #535  which implements changes discussed above.

**Comment by jscanvic** [comment]
Should this be closed? @tachella
